File: code_samples/ruby/performance_memoization.rb
# Performance: Memoization
# AI/ML Training Sample

class Memoization
  attr_accessor :data
  
  def initialize
    @data = ""
  end
  
  def process(input)
    @data = input
  end
  
  def validate
    !@data.empty?
  end
  
  def to_s
    "Data: #{@data}"
  end
end

# Example usage
instance = Memoization.new
instance.process("example")
puts instance
puts "Valid: #{instance.validate}"

File: code_samples/r/performance_memoization.r
# Performance: Memoization
# AI/ML Training Sample

Memoization <- setRefClass(
    "Memoization",
    fields = list(data = "character"),
    methods = list(
        initialize = function() {
            data <<- ""
        },
        process = function(input) {
            data <<- input
        },
        getData = function() {
            return(data)
        },
        validate = function() {
            return(nchar(data) > 0)
        }
    )
)

# Example usage
instance <- Memoization$new()
instance$process("example")
cat("Data:", instance$getData(), "\n")
cat("Valid:", instance$validate(), "\n")
